src/utils/timer.ts

```diff
@@ -0,0 +1,74 @@
+import { logWarn } from '.';
+
+interface TimerProps {
+    /** Set a upper threshold for the timer, which will make the timer autostop. Default is 2000 milliseconds. */
+    maxTime?: number;
+    /** Optional callback. Check the this reference binding as it will default to {Timer} */
+    callback?: () => void;
+}
+
+type TimerStatus = 'Started' | 'Stopped' | 'Over time' | 'Not started';
+
+export class Timer {
+    /** Time elapsed is stored as milliseconds. */
+    private _timeElapsed = 0;
+    private _startTime?: Date;
+
+    /** An optional callback when the timer stops. */
+    private _callback?: () => void;
+
+    /**
+     * We use a max time for the timer to avoid it counting "forever".
+     * This value is configurable during construction.
+     */
+    private _maxTime: number;
+    private _timerId?: NodeJS.Timeout;
+    private _status: TimerStatus;
+
+    constructor(props?: TimerProps) {
+        if (props && props.maxTime && typeof props.maxTime === 'number' && !isNaN(props.maxTime)) {
+            this._maxTime = props.maxTime;
+        } else this._maxTime = 2000;
+
+        this._status = 'Not started';
+    }
+
+    /** Returns the current elapsed time of the timer. Don't use this number for very time-critical use-cases. */
+    public get timeElapsed(): number {
+        if (this._startTime) {
+            return Date.now() - this._startTime?.getTime();
```

Comment:
After it has been stopped, it should return the same timeElapsed as when it was stopped? Now it continues counting after it was stopped.

Reply:
Thanks, I have introduced finalTime for use in recording the final time. timeElapsed is now the time when the timer is running, and roughly equal to finalTime once the timer is stopped.